v4.0.x: Re-add removed deprecate-only MPI-2.0 symbols

gpaulsen opened the pull request:
I restored the APIs in mpi.h in commit 2d3b4bb9, but forgot the Makefile changes.  :(
 
Fixes #6114

Signed-off-by: Bert Wesarg <bert.wesarg@tu-dresden.de>

gpaulsen commented:
Thanks I'll take a look

ggouaillardet commented:
is there any reason why this is not a cherry-pick/backport of #6118 ?

ggouaillardet commented:
you will also need to cherry-pick the commit from #6122 once merged into master

bertwesarg commented:
BTW, if it comes to the release time, I don't think that this fix should effect the shared library version at all. As 4.0.0 broke the `c:r:a` anyway.

gpaulsen commented:
Okay, previous CI passed on two commits on this branch. 

I've now pushed the 3rd and final commit to this branch, which restores the removed functions to the mpi library (but not to the header, unless the configure flag is thrown).

@jsquyres, @hjelmn please re-review.

jsquyres commented on ompi/mpi/c/Makefile.am:
```diff
@@ -437,7 +437,8 @@ libmpi_c_mpi_la_SOURCES = \
         win_wait.c
 
 
-if OMPI_ENABLE_MPI1_COMPAT
```
How about putting these files just back up (alphabetically) in the main list of files?

Also, this specific commit was not a cherry pick, right?  Can you please put a comment about that in the commit message? (i.e., why this is not a cherry pick: because on master, we are no longer building those symbols by default, but on v4.0 we *are* still building those symbols by default, yadda yadda yadda)

gpaulsen replied:
I kinda liked having them separated out at the bottom to help indicate that they're going away.

I will rework the commit message to indicate why we want/need this v4.0.x specific patch (instead of a cherry-pick from master).

gpaulsen commented:
Drat, I just saw that too... working.

ibm-ompi commented:
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/fe832eba609631aa95f53983e34de01d

ibm-ompi commented:
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/30c8322f19b427bcad29721be9a5e53c

ibm-ompi commented:
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/4f23e242e8c62b305455a39c368a37f4

ibm-ompi commented:
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/eba77f14eb4c28e794b806274c75ee62

bertwesarg commented:
I think `OMPI_OMIT_MPI1_COMPAT_DECLS` must be 0 if `OMPI_BUILDING` is 1. But it looks like it's not:

```c
#if !defined(OMPI_OMIT_MPI1_COMPAT_DECLS)
#    define OMPI_OMIT_MPI1_COMPAT_DECLS !OMPI_ENABLE_MPI1_COMPAT
#endif
```

ibm-ompi commented:
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/062b323ebc9bef2b179451e13b8c8a06

gpaulsen commented:
Okay, I just repushed.  And instead in the Fortran implementations I set OMPI_OMIT_MPI1_COMPAT_DECLS to 0.  Isn't that what we want, instead of !OMPI_ENABLE_MPI1_COMPAT ?

ggouaillardet commented:
@bertwesarg you can already get rid of the deprecation warning by adding `-DOMPI_WANT_MPI_INTERFACE_WARNING=0` to your `CPPFLAGS`. At first glance, this PR should not change this.

bertwesarg commented:
> @bertwesarg you can already get rid of the deprecation warning by adding `-DOMPI_WANT_MPI_INTERFACE_WARNING=0` to your `CPPFLAGS`. At first glance, this PR should not change this.

Stupid me, we already do this. I think I messed this up with deprecated SHMEM symbols. But not necessarily OSHMEM in Open MPI.

ggouaillardet commented:
@bertwesarg  if you did not `configure --enable-mpi1-compatibility`, you can use the removed symbols in `mpi.h` by adding `-DOMPI_OMIT_MPI1_COMPAT_DECLS=0` to your `CPPFLAGS`.

At first glance, I was a bit surprised that with this flag only, there is no warning we are using a removed symbol (e.g. `MPI_Type_lb has been removed in MPI-3.0; use MPI_Type_get_extent instead.`).
Is this the expected behavior ?

ggouaillardet commented:
Incidentally, we might want to revamp the logic on `master`. Since the removed symbols are not built by default, we should not allow the enduser to "restore" them in the header files via some macros.

bertwesarg commented:
> Incidentally, we might want to revamp the logic on `master`. Since the removed symbols are not built by default, we should not allow the enduser to "restore" them in the header files via some macros.

Yes, it makes only sense to restore the prototypes, if the symbols are actual available in the library. Thus we shouldn't  misuse `OMPI_OMIT_MPI1_COMPAT_DECLS` here.

gpaulsen commented:
I've rebased my branch on top of v4.0.x, and modified the commit message of the non-cherry-picked commit to include:
```
    NOTE: This commit is *NOT* a cherry-pick from master, because
          on master, we are no longer building those symbols by
          default, but on v4.0.x we _ARE_ still building these
          symbols by default.   This is because the v4.0.x branch
          is to remain backwards compatible with v3.0.x, while at
          the same time removing the "removed" symbols from mpi.h
          (unless the user configures with --enable-mpi1-compatibility)
```

I did not alphabetize the removed implementation filenames in the Makefile.am, as this looks cleaner and more obvious of our intent of removing these in the future (in my opinion).

jsquyres commented:
...we're not *ever* going to remove them here on the v4.0.x branch, right?

gpaulsen commented:
:(  I'll take a look

gpaulsen commented:
I've got a fix, and will push soon.

gpaulsen commented:
@jsquyres please review.  I can squish the last two commits if you'd like.

bertwesarg commented on ompi/mpi/c/address.c:
```diff
@@ -21,6 +21,17 @@
+ * to the newer MPI calls, we are defining OMPI_OMIT_MPI1_COMPAT_DECLS
+ * to 0 before including the c bindings.
+ */
+#define OMPI_OMIT_MPI1_COMPAT_DECLS 0
```
Shouldn't the `OMPI_BUILDING` `#define` be sufficient here? If not, why?

gpaulsen replied:
So, there's currently no logic to affect `OMPI_OMIT_MPI1_COMPAT_DECLS` via `OMPI_BUILDING`.  Are you proposing we add that?

bertwesarg replied:
`OMPI_BUILDING` already overrules `OMPI_WANT_MPI_INTERFACE_WARNING`, where the latter is for users. Why shouldn't `OMPI_BUILDING` also overrule `OMPI_OMIT_MPI1_COMPAT_DECLS`, and leave `OMPI_OMIT_MPI1_COMPAT_DECLS` to the users?

gpaulsen replied:
What do you mean by _"leave `OMPI_OMIT_MPI1_COMPAT_DECLS` to the users?"_  Users can control that already via the --enable-mpi-compatibility configure flag.  This is just silently unsetting during the build of a few files in the .c files (not mpi.h) to allow those files to build.

bertwesarg replied:
As long as the symbols are in the library, I think its valid to access them by there prototype. I can understand that by default you hide them, but having a way to reveal the prototypes is very convenient. As not all users can wait until the used software project changes the software to not use deprecated APIs.

As a second _user case_, we the tools developers nevertheless want to wrap as many API as possible. If we know, that if we set some magic variable and get all prototypes for all available symbols. We certainly will use it.

If `OMPI_OMIT_MPI1_COMPAT_DECLS` is that hook, to get all prototypes, than its something for the _users_.

jsquyres replied:
@gpaulsen I think the point here is that `OMPI_BUILDING` is a flag we use internally -- it means that additional things should be available because we are building Open MPI itself (vs. a user MPI application).  I think @bertwesarg is saying that `OMPI_BUILDING` should probably *also* imply that the MPI1 compatibility prototypes should be available if `OMPI_BUILDING==1`.  That way, `OMPI_OMIT_MPI1_COMPAT_DECLS` is purely tied to `--enable-mpi-compatibility` (i.e., solely "for the users").

gpaulsen replied:
I see, let me see if I can do that and repush, for your review.

gpaulsen commented on ompi/mpi/c/address.c:
```diff
@@ -21,6 +21,17 @@
+ *
+ * To prevent having to port these implementations of removed functions
+ * to the newer MPI calls, we are defining ENABLE_MPI1_COMPAT to 1
+ * before including the c bindings.
```
The intent was to communicate that we are NOT porting the implementations of these removed mpi1 functions to use the NEWer APIs, and instead we're doing Maco magic to allow us to build the older functions for the library only.
This latest push changed the mechanism that we do this, and no longer do we SET ENABLE_MPI1_COMPAT internally when we build these few routines, but instead this PR adds a `|| OMPI_BUILDING` to the cpp logic around the prototypes of these removed functions to allow us to build them into the library.

I suspect all of this is clear ( `|| OMPI_BUILDING` is much more straight forward), so I've just removed this paragraph.

hppritcha commented:
I'm getting tire of this .  @jsquyres we are going to take this as is.

gpaulsen commented:
Sorry, I put work in progress on here because There was a real change that was still needed.  I'll add another commit on top of this in a separate PR.

hppritcha commented:
oops I didn't notice the DWI tag.  @gpaulsen revert an reopen if there was more to go in.  sorry.  shouldn't do these with my iphone.

jsquyres commented:
This PR should be reverted.  It was not ready to be merged.

gpaulsen commented:
Okay, I can revert this.
Do I then open another PR for the revert, and another for the final?

jsquyres commented:
Just click the "revert" button on this PR.

bertwesarg commented:
> @bertwesarg if you did not `configure --enable-mpi1-compatibility`, you can use the removed symbols in `mpi.h` by adding `-DOMPI_OMIT_MPI1_COMPAT_DECLS=0` to your `CPPFLAGS`.

This does not work with the 4.0.1rc1, as I only get a

```
include/mpi.h:315:0: warning: "OMPI_OMIT_MPI1_COMPAT_DECLS" redefined
```

but not the hidden prototypes. The following patch to v4.0.1 gives me the desired behavior:

```diff
diff --git i/ompi/include/mpi.h.in w/ompi/include/mpi.h.in
index 27feae4ad8..03f095013b 100644
--- i/ompi/include/mpi.h.in
+++ w/ompi/include/mpi.h.in
@@ -306,7 +306,8 @@
           * Don't do MACRO magic for building Profiling library as it
           * interferes with the above.
           */
-#        if (OMPI_ENABLE_MPI1_COMPAT || OMPI_BUILDING)
+#        if defined(OMPI_OMIT_MPI1_COMPAT_DECLS)
+#        elif (OMPI_ENABLE_MPI1_COMPAT || OMPI_BUILDING)
 #            define OMPI_OMIT_MPI1_COMPAT_DECLS 0
 #            define OMPI_REMOVED_USE_STATIC_ASSERT 0
 #            define __mpi_interface_removed__(func, newfunc) __mpi_interface_deprecated__(#func " was removed in MPI-3.0.  Use " #newfunc " instead.  continuing...")
```
